File: codes/HellDream/63.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int rows = obstacleGrid.length;
        int cols = obstacleGrid[0].length;
        int[][] paths = new int[rows][cols];
        for(int col=0;col<cols;col++){
            if(obstacleGrid[0][col]==1) break;
            paths[0][col] = 1;
        }
        for(int row=0;row<rows;row++){
            if(obstacleGrid[row][0]==1) break;
            paths[row][0] = 1;
        }
        
        for(int row=1;row<rows;row++){
            for(int col=1;col<cols;col++){
                if(obstacleGrid[row][col]==1){
                    paths[row][col] = 0;
                }else{
                    paths[row][col] = paths[row-1][col]+paths[row][col-1];
                }
            }
        }
        return paths[rows-1][cols-1];
    }
}

File: codes/Garnetwzy/63.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        if(m == 0)
            return 0;
        int n  = obstacleGrid[0].size();
        if(obstacleGrid[0][0] == 1)
            return 0;
        vector<vector<long>> dp(m, vector<long>(n, 0));
        dp[0][0] = 1;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(i == 0 && j == 0)
                    continue;
                if(obstacleGrid[i][j] == 1) {
                    dp[i][j] = 0;
                    continue;
                }
                int first = -1;
                int second = -1;
                if(j > 0 && obstacleGrid[i][j-1] != 1) {
                    first = dp[i][j-1];
                }
                if(i > 0 && obstacleGrid[i-1][j] != 1) {
                    second = dp[i-1][j];
                }
                dp[i][j] = (long)max(first, 0) + (long)max(second, 0);
            }
        }
        return dp[m-1][n-1];
    }
};